mlir/lib/Dialect/Rock/Transforms/ReuseLDS.cpp

```diff
@@ -0,0 +1,355 @@
+  // Check for arch limitations exceeded
+  FailureOr<StringAttr> maybeArch = getArch(op);
+  if (succeeded(maybeArch)) {
+    StringAttr arch = maybeArch.value();
+    const int64_t ldsSize = rock::lookupArchInfo(arch).maxSharedMemPerWG;
+    return success(ldsBytes <= ldsSize);
+  }
+  return success();
+}
+
+static int64_t
+assignColors(GpuAllocOp alloc, llvm::SetVector<int64_t> &usedColors,
+             llvm::MapVector<int64_t, int64_t> &colorMemSize,
+             llvm::MapVector<GpuAllocOp, SetVector<int64_t>> &colorAssignment) {
+  const int64_t requiredSize =
+      getWorkgroupMemorySize(alloc.getOutput().getType());
+  assert(requiredSize > 0);
+  int64_t color = 0;
+  int64_t allocatedSize = 0;
+  int64_t maxColor = 0;
+
+  while (allocatedSize < requiredSize) {
+    if (usedColors.contains(color)) {
+      // found a used color, all the assigned colors must be consecutive
+      // let's start again
+      allocatedSize = 0;
+      colorAssignment[alloc].clear();
+      // Find the first available color
+      while (usedColors.contains(color)) {
+        color++;
+      }
+    }
+    maxColor = std::max(maxColor, color);
+    colorAssignment[alloc].insert(color);
+    bool existingColor = colorMemSize.contains(color);
+    // New color
+    if (!existingColor) {
+      // Make this aligned to 128 bits
+      colorMemSize[color] = roundUpToMultipleOf16(requiredSize - allocatedSize);
+    }
+    allocatedSize += colorMemSize[color];
+    color++;
+  }
+  return maxColor;
+}
+
+static std::tuple<int64_t, SmallVector<std::tuple<GpuAllocOp, int64_t, bool>>>
+graphColoring(
+    SmallVector<GpuAllocOp> &gpuAllocs,
+    llvm::MapVector<GpuAllocOp, SetVector<GpuAllocOp>> &interferenceGraph) {
+  llvm::MapVector<GpuAllocOp, SetVector<int64_t>> colorAssignment;
+  llvm::MapVector<int64_t, int64_t> colorMemSize;
+  int64_t maxColor = 0;
+
+  SmallVector<GpuAllocOp> sortedAllocs(gpuAllocs);
+
+  // Sort by alloc size
+  llvm::sort(sortedAllocs, [](GpuAllocOp &a, GpuAllocOp &b) {
+    auto aSize = getWorkgroupMemorySize(a.getOutput().getType());
+    auto bSize = getWorkgroupMemorySize(b.getOutput().getType());
+    return aSize < bSize;
+  });
+  // Assign colors using greedy algorithm
+  for (GpuAllocOp alloc : sortedAllocs) {
+    llvm::SetVector<int64_t> usedColors;
+    for (GpuAllocOp neighbor : interferenceGraph[alloc]) {
+      if (colorAssignment.find(neighbor) != colorAssignment.end()) {
+        for (int64_t color : colorAssignment[neighbor]) {
+          usedColors.insert(color);
+        }
+      }
+    }
+
+    // Assign a set of colors
+    int64_t localMaxColor =
+        assignColors(alloc, usedColors, colorMemSize, colorAssignment);
+    maxColor = std::max(maxColor, localMaxColor);
+  }
+
+  // Compute required memory and offsets per color
+  SmallVector<int64_t> colorOffsets;
+  int64_t offset = 0;
+  for (int64_t color = 0; color < maxColor + 1; color++) {
```

Comment:
Could just loop through `colorMemSize`, maybe `llvm::enumerate`'d. Or use some sort of iota?